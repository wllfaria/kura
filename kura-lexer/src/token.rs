#[allow(clippy::module_inception)]
mod token;

mod kind;
mod operator;
mod primitive;
mod value;

pub use kind::Kind;
pub use operator::Operator;
pub use primitive::{FloatSizes, IntSizes, Primitive, UIntSizes};
pub use token::{IntoToken, Location, Token};
pub use value::Value;
