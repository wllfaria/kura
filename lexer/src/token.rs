#[derive(Debug)]
pub enum NumSize {
    F8,
    F16,
    F32,
    F64,
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    Usize,
    Isize,
}

impl std::fmt::Display for NumSize {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            NumSize::F8 => write!(f, "f8"),
            NumSize::F16 => write!(f, "f16"),
            NumSize::F32 => write!(f, "f32"),
            NumSize::F64 => write!(f, "f64"),
            NumSize::I8 => write!(f, "i8"),
            NumSize::I16 => write!(f, "i16"),
            NumSize::I32 => write!(f, "i32"),
            NumSize::I64 => write!(f, "i64"),
            NumSize::U8 => write!(f, "u8"),
            NumSize::U16 => write!(f, "u16"),
            NumSize::U32 => write!(f, "u32"),
            NumSize::U64 => write!(f, "u64"),
            NumSize::Usize => write!(f, "usize"),
            NumSize::Isize => write!(f, "isize"),
        }
    }
}

impl TryFrom<&str> for NumSize {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(NumSize::I8),
            "i16" => Ok(NumSize::I16),
            "i32" => Ok(NumSize::I32),
            "i64" => Ok(NumSize::I64),
            "u8" => Ok(NumSize::U8),
            "u16" => Ok(NumSize::U16),
            "u32" => Ok(NumSize::U32),
            "u64" => Ok(NumSize::U64),
            "f8" => Ok(NumSize::F8),
            "f16" => Ok(NumSize::F16),
            "f32" => Ok(NumSize::F32),
            "f64" => Ok(NumSize::F64),
            "usize" => Ok(NumSize::Usize),
            "isize" => Ok(NumSize::Isize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Token<'tok> {
    Integer { value: u64, size: Option<NumSize> },
    SignedInteger { value: i64, size: Option<NumSize> },
    Float { value: f64, size: Option<NumSize> },
    Minus,
    String(&'tok str),
}

impl std::fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Integer { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::SignedInteger { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Minus => write!(f, "-"),
            Token::String(s) => write!(f, "{s}"),
        }
    }
}
