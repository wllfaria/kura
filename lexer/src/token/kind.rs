use miette::Error;

use super::{
    operator::Operator,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum Kind<'tok> {
    Value(Value<'tok>),
    Op(Operator),

    Var,
    Const,
    Match,
    If,
    Else,
    Fun,
    Struct,
    Enum,
    Return,
    Eof,
}

impl<'tok> Kind<'tok> {
    pub fn infix_precedence(&self) -> Result<(u8, u8), Error> {
        match self {
            Kind::Op(Operator::Plus) | Kind::Op(Operator::Minus) => Ok((1, 2)),
            Kind::Op(Operator::Less)
            | Kind::Op(Operator::LessEqual)
            | Kind::Op(Operator::Greater)
            | Kind::Op(Operator::GreaterEqual) => Ok((3, 4)),
            Kind::Op(Operator::Star) | Kind::Op(Operator::Slash) => Ok((5, 6)),
            _ => miette::bail!("invalid operator with no precedence"),
        }
    }

    pub fn is_binary_op(&self) -> bool {
        matches!(
            self,
            Kind::Op(Operator::Star)
                | Kind::Op(Operator::Slash)
                | Kind::Op(Operator::Plus)
                | Kind::Op(Operator::Minus)
                | Kind::Op(Operator::Greater)
                | Kind::Op(Operator::Less)
                | Kind::Op(Operator::LessEqual)
                | Kind::Op(Operator::GreaterEqual)
        )
    }

    pub fn identifier_from(value: &'tok str) -> Kind<'tok> {
        match value {
            "var" => Kind::Var,
            "const" => Kind::Const,
            "match" => Kind::Match,
            "if" => Kind::If,
            "else" => Kind::Else,
            "fun" => Kind::Fun,
            "struct" => Kind::Struct,
            "enum" => Kind::Enum,
            "return" => Kind::Return,
            _ => Kind::Value(Value::Ident(value)),
        }
    }
}

impl<'tok> IntoToken<'tok> for Kind<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(self, (start_byte, end_byte).into())
    }
}
