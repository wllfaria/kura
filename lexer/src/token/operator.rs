use std::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
};

#[derive(Debug)]
pub enum Operator {
    LeftParen,
    RightParen,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    MinusEqual,
    Plus,
    PlusEqual,
    Equal,
    ThickArrow,
    EqualEqual,
    Star,
    StarEqual,
    Ampersand,
    Slash,
    SlashEqual,
    Colon,
    SemiColon,
    Bang,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    NotEqual,
    And,
    Or,
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Operator::Minus => write!(f, "-"),
            Operator::MinusEqual => write!(f, "-="),
            Operator::Colon => write!(f, ":"),
            Operator::SemiColon => write!(f, ";"),
            Operator::LeftParen => write!(f, "("),
            Operator::RightParen => write!(f, ")"),
            Operator::LeftBracket => write!(f, "["),
            Operator::RightBracket => write!(f, "]"),
            Operator::LeftBrace => write!(f, "{{"),
            Operator::RightBrace => write!(f, "}}"),
            Operator::Comma => write!(f, ","),
            Operator::Dot => write!(f, "."),
            Operator::Plus => write!(f, "+"),
            Operator::PlusEqual => write!(f, "+="),
            Operator::Equal => write!(f, "="),
            Operator::ThickArrow => write!(f, "=>"),
            Operator::EqualEqual => write!(f, "=="),
            Operator::Star => write!(f, "*"),
            Operator::StarEqual => write!(f, "*="),
            Operator::Ampersand => write!(f, "&"),
            Operator::Slash => write!(f, "/"),
            Operator::Bang => write!(f, "!"),
            Operator::Less => write!(f, "<"),
            Operator::LessEqual => write!(f, "<="),
            Operator::Greater => write!(f, ">"),
            Operator::GreaterEqual => write!(f, ">="),
            Operator::NotEqual => write!(f, "!="),
            Operator::SlashEqual => write!(f, "/="),
            Operator::And => write!(f, "&&"),
            Operator::Or => write!(f, "||"),
        }
    }
}

impl<'tok> IntoToken<'tok> for Operator {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(Kind::Op(self), (start_byte, end_byte).into())
    }
}
