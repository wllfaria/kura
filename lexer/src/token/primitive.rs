use core::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum FloatSizes {
    F8,
    F16,
    F32,
    F64,
}

impl fmt::Display for FloatSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FloatSizes::F8 => write!(f, "f8"),
            FloatSizes::F16 => write!(f, "f16"),
            FloatSizes::F32 => write!(f, "f32"),
            FloatSizes::F64 => write!(f, "f64"),
        }
    }
}

#[derive(Debug)]
pub enum IntSizes {
    I8,
    I16,
    I32,
    I64,
    Isize,
}

impl fmt::Display for IntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            IntSizes::I8 => write!(f, "i8"),
            IntSizes::I16 => write!(f, "i16"),
            IntSizes::I32 => write!(f, "i32"),
            IntSizes::I64 => write!(f, "i64"),
            IntSizes::Isize => write!(f, "isize"),
        }
    }
}

#[derive(Debug)]
pub enum UIntSizes {
    U8,
    U16,
    U32,
    U64,
    Usize,
}

impl fmt::Display for UIntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UIntSizes::U8 => write!(f, "u8"),
            UIntSizes::U16 => write!(f, "u16"),
            UIntSizes::U32 => write!(f, "u32"),
            UIntSizes::U64 => write!(f, "u64"),
            UIntSizes::Usize => write!(f, "usize"),
        }
    }
}

impl TryFrom<&str> for IntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(IntSizes::I8),
            "i16" => Ok(IntSizes::I16),
            "i32" => Ok(IntSizes::I32),
            "i64" => Ok(IntSizes::I64),
            "isize" => Ok(IntSizes::Isize),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for FloatSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "f8" => Ok(FloatSizes::F8),
            "f16" => Ok(FloatSizes::F16),
            "f32" => Ok(FloatSizes::F32),
            "f64" => Ok(FloatSizes::F64),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for UIntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "u8" => Ok(UIntSizes::U8),
            "u16" => Ok(UIntSizes::U16),
            "u32" => Ok(UIntSizes::U32),
            "u64" => Ok(UIntSizes::U64),
            "usize" => Ok(UIntSizes::Usize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Primitive {
    Bool(bool),
    UInt {
        value: u64,
        size: Option<UIntSizes>,
    },
    Int {
        value: i64,
        size: Option<IntSizes>,
    },
    Float {
        value: f64,
        size: Option<FloatSizes>,
    },
}

impl fmt::Display for Primitive {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Primitive::Bool(b) => write!(f, "{b}"),
            Primitive::UInt { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Int { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
        }
    }
}

impl<'tok> IntoToken<'tok> for Primitive {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(
            Kind::Value(Value::Primitive(self)),
            (start_byte, end_byte).into(),
        )
    }
}
