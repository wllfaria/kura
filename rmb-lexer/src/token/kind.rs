use miette::Error;

use super::{
    operator::Operator,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug, PartialEq)]
pub enum Kind<'tok> {
    Value(Value<'tok>),
    Op(Operator),

    Var,
    Const,
    Match,
    If,
    Else,
    Fun,
    Struct,
    Enum,
    Return,
    Eof,
}

impl std::fmt::Display for Kind<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Kind::Value(val) => write!(f, "{}", val),
            Kind::Op(op) => write!(f, "{}", op),
            Kind::Var => write!(f, "var"),
            Kind::Const => write!(f, "const"),
            Kind::Match => write!(f, "match"),
            Kind::If => write!(f, "if"),
            Kind::Else => write!(f, "else"),
            Kind::Fun => write!(f, "fun"),
            Kind::Struct => write!(f, "struct"),
            Kind::Enum => write!(f, "enum"),
            Kind::Return => write!(f, "return"),
            Kind::Eof => write!(f, "eof"),
        }
    }
}

impl<'tok> Kind<'tok> {
    pub fn infix_precedence(&self) -> Result<(u8, u8), Error> {
        match self {
            Kind::Op(Operator::Plus) | Kind::Op(Operator::Minus) => Ok((1, 2)),
            Kind::Op(Operator::Less)
            | Kind::Op(Operator::LessEqual)
            | Kind::Op(Operator::EqualEqual)
            | Kind::Op(Operator::Or)
            | Kind::Op(Operator::Greater)
            | Kind::Op(Operator::GreaterEqual) => Ok((3, 4)),
            Kind::Op(Operator::Star) | Kind::Op(Operator::Slash) => Ok((5, 6)),
            _ => miette::bail!("invalid operator with no precedence"),
        }
    }

    pub fn is_binary_op(&self) -> bool {
        matches!(
            self,
            Kind::Op(Operator::Star)
                | Kind::Op(Operator::Slash)
                | Kind::Op(Operator::And)
                | Kind::Op(Operator::Or)
                | Kind::Op(Operator::Plus)
                | Kind::Op(Operator::Minus)
                | Kind::Op(Operator::Greater)
                | Kind::Op(Operator::EqualEqual)
                | Kind::Op(Operator::Less)
                | Kind::Op(Operator::LessEqual)
                | Kind::Op(Operator::NotEqual)
                | Kind::Op(Operator::GreaterEqual)
        )
    }

    pub fn identifier_from(value: &'tok str) -> Kind<'tok> {
        match value {
            "var" => Kind::Var,
            "const" => Kind::Const,
            "match" => Kind::Match,
            "if" => Kind::If,
            "else" => Kind::Else,
            "fun" => Kind::Fun,
            "struct" => Kind::Struct,
            "enum" => Kind::Enum,
            "return" => Kind::Return,
            _ => Kind::Value(Value::Ident(value)),
        }
    }
}

impl<'tok> IntoToken<'tok> for Kind<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(self, (start_byte, end_byte).into())
    }
}
