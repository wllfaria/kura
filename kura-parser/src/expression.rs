use kura_lexer::token::{Kind, Location, Operator, Primitive, Token, Value};
use kura_lexer::{Lexer, TransposeRef};

use crate::Expression;

mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 3;
    pub const MUL: u8 = 4;
    pub const ASSOC: u8 = 5;
    pub const APPLY: u8 = 6;
}

fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::Star | Operator::Slash => precedences::MUL,
        Operator::And => precedences::ASSOC,
        Operator::LeftParen | Operator::EqualEqual | Operator::NotEqual => precedences::APPLY,
        _ => precedences::BASE,
    }
}

pub fn parse_expression<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    match lexer.peek().transpose().map_err(ToString::to_string)? {
        Some(token) => match token.kind {
            Kind::Var | Kind::Const => parse_variable(lexer),
            _ => parse_with_precedence(lexer, precedences::BASE, base_expr),
        },
        None => unreachable!(),
    }
}

pub fn parse_identifier<'parser>(lexer: &mut Lexer<'parser>) -> Result<(Expression<'parser>, &'parser str), String> {
    let name_and_loc = lexer
        .next()
        .transpose()
        .map_err(|e| e.to_string())?
        .map(|token| match token.kind {
            Kind::Value(Value::Ident(name)) => (name, token.location),
            _ => ("", token.location),
        });

    match name_and_loc {
        Some(("", location)) => Err(location.to_string()),
        Some((name, location)) => Ok((Expression::Ident { name, location }, name)),
        None => Err("TODO".into()),
    }
}

fn parse_expr_block<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    let mut expressions = vec![];

    let block_start = lexer.expect(Kind::Op(Operator::LeftBrace)).map_err(|e| e.to_string())?;

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightBrace)) => {
                break;
            }
            None => break,
            _ => (),
        }

        let expr = parse_expression(lexer, base_expr)?;
        expressions.push(expr);
    }

    let block_end = lexer
        .expect(Kind::Op(Operator::RightBrace))
        .map_err(|e| e.to_string())?;

    let location = block_start.location.start_byte..block_end.location.end_byte;
    Ok(Expression::Block {
        expressions,
        location: location.into(),
    })
}

fn parse_variable<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer
        .expect_one_of(&[Kind::Var, Kind::Const])
        .map_err(|e| e.to_string())?;
    let mutable = matches!(keyword.kind, Kind::Var);

    let (_, name) = parse_identifier(lexer)?;

    let typ = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::Colon)) => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            Some(parse_identifier(lexer)?)
        }
        _ => None,
    };

    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match token.kind {
            Kind::Op(Operator::LeftBrace) => parse_expr_block(lexer, false)?,
            _ => parse_expression(lexer, false)?,
        },
        _ => unreachable!(),
    };

    lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = Location::new(keyword.location.start_byte, value.location().end_byte);
    Ok(Expression::Var {
        mutable,
        typ: typ.map(|(typ, _)| Box::new(typ)),
        name,
        value: Box::new(value),
        location,
    })
}

fn parse_if_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::If).map_err(|e| e.to_string())?;

    let condition = parse_expression(lexer, false)?;

    let body = parse_expr_block(lexer, true)?;

    let mut falsy_branches = vec![];

    while let Some(keyword) = lexer.peek().transpose().map_err(|e| e.to_string())? {
        if keyword.kind == Kind::Else {
            // consume `else`
            lexer.next().transpose().map_err(|e| e.to_string())?;

            if lexer
                .peek()
                .transpose()
                .map_err(|e| e.to_string())?
                .map(|t| matches!(t.kind, Kind::If))
                .unwrap_or(false)
            {
                let else_if = parse_if_expression(lexer)?;
                falsy_branches.push(else_if);
            } else {
                let else_block = parse_expr_block(lexer, true)?;
                falsy_branches.push(else_block);
                break;
            }
        } else {
            break;
        }
    }

    let end = match falsy_branches.last() {
        Some(branch) => branch.location().end_byte,
        None => body.location().end_byte,
    };

    let location = keyword.location.start_byte..end;
    Ok(Expression::If {
        condition: Box::new(condition),
        truthy: Box::new(body),
        falsy: falsy_branches,
        location: location.into(),
    })
}

fn parse_value<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Value(value) = kind else {
        unreachable!();
    };

    match value {
        Value::Primitive(_) => parse_primitive(lexer),
        Value::Ident(_) => Ok(parse_identifier(lexer)?.0),
        t => todo!("{t}"),
    }
}

fn parse_operation<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Op(op) = kind else {
        unreachable!();
    };

    match op {
        Operator::LeftParen => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            let left = parse_with_precedence(lexer, precedences::BASE, false)?;
            lexer
                .expect(Kind::Op(Operator::RightParen))
                .map_err(|e| e.to_string())?;
            Ok(left)
        }
        Operator::LeftBrace => parse_expr_block(lexer, true),
        t => todo!("{t:?}"),
    }
}

fn parse_fun_call<'parser>(
    lexer: &mut Lexer<'parser>,
    ident: Expression<'parser>,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::LeftParen)).map_err(|e| e.to_string())?;

    let mut arguments = vec![];

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightParen)) => {
                break;
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Comma)) => {
                lexer.next().transpose().map_err(|e| e.to_string())?;
                continue;
            }
            None => break,
            _ => (),
        }

        let arg = parse_expression(lexer, false)?;

        arguments.push(arg);
    }

    let close_paren = lexer
        .expect(Kind::Op(Operator::RightParen))
        .map_err(|e| e.to_string())?;

    if base_expr {
        lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;
    }

    let location = ident.location().start_byte..close_paren.location.end_byte;
    let expr = Expression::FunCall {
        ident: Box::new(ident),
        location: location.into(),
        arguments,
    };
    Ok(expr)
}

fn parse_assign<'parser>(lexer: &mut Lexer<'parser>, left: Expression<'parser>) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::LeftBrace)) => parse_expr_block(lexer, false)?,
        Some(_) => parse_expression(lexer, false)?,
        _ => unreachable!(),
    };

    let closing = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = left.location().start_byte..closing.location.end_byte;

    Ok(Expression::Assign {
        ident: Box::new(left),
        location: location.into(),
        value: Box::new(value),
    })
}

fn parse_with_precedence<'parser>(
    lexer: &mut Lexer<'parser>,
    min_precedence: u8,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    let mut left = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match &token.kind {
            Kind::Value(_) => parse_value(lexer)?,
            Kind::Op(_) => parse_operation(lexer)?,
            Kind::Return => parse_return_expression(lexer)?,
            Kind::If => parse_if_expression(lexer)?,
            t => todo!("{t:?}"),
        },
        None => todo!(),
    };

    if let Expression::Ident { .. } = left {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::LeftParen)) => {
                return parse_fun_call(lexer, left, base_expr);
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Equal)) => return parse_assign(lexer, left),
            _ => (),
        }
    }

    loop {
        let Some(next) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
            return Ok(left);
        };

        let Kind::Op(operator) = next.kind else {
            return Ok(left);
        };

        if !next.kind.is_binary_op() {
            return Ok(left);
        }

        let precedence = get_precedence(operator);

        if precedence <= min_precedence {
            break;
        }

        let Some(_) = lexer.next().transpose().map_err(|e| e.to_string())? else {
            unreachable!();
        };

        let right = parse_with_precedence(lexer, precedence, false)?;

        let location = Location::new(left.location().start_byte, right.location().end_byte);
        left = Expression::BinaryOp {
            lhs: Box::new(left),
            operator,
            rhs: Box::new(right),
            location,
        };
    }

    Ok(left)
}

fn parse_return_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::Return).map_err(|e| e.to_string())?;

    let value = parse_expression(lexer, false)?;

    let ending_semi = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = keyword.location.start_byte..ending_semi.location.end_byte;
    Ok(Expression::Return {
        value: Box::new(value),
        location: location.into(),
    })
}

fn parse_primitive<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let (primitive, location) = match lexer.next().transpose().map_err(|e| e.to_string())? {
        Some(Token {
            kind: Kind::Value(Value::Primitive(primitive)),
            location,
            ..
        }) => (primitive, location),
        _ => unreachable!(),
    };

    match primitive {
        Primitive::Int { value, size } => Ok(Expression::IntLiteral { value, size, location }),
        Primitive::UInt { value, size } => Ok(Expression::UintLiteral { value, size, location }),
        Primitive::Float { value, size } => Ok(Expression::FloatLiteral { value, size, location }),
        Primitive::Bool(value) => Ok(Expression::Bool { value, location }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Parser;

    fn make_sut(source: &str) -> Parser<'_> {
        let lexer = Lexer::new(source);
        Parser::new(source, lexer)
    }

    #[test]
    fn simple_math() {
        let simple_math_expr = "1 + 2 * 3 - 4";

        let mut parser = make_sut(simple_math_expr);

        let math_expr_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(math_expr_ast);
    }

    #[test]
    fn mutable_variable() {
        let variables = "var hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn immutable_constant() {
        let variables = "const hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn if_statement() {
        let source = r#"
            if something == another_thing && 1 != 2 {
                const this_is_a_var = 10 + 3;
            } else if something == 10 {
                const this_is_another = 10 + 10;
            } else {
                const omg = 1 + 1;
            }
        "#;
        let mut parser = make_sut(source);

        let if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(if_ast);
    }

    #[test]
    fn if_as_variable_value() {
        let source = r#"
            const my_var = if something == another_thing {
                10 + 10
            } else {
                20 + 20
            };
        "#;

        let mut parser = make_sut(source);
        let let_if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(let_if_ast);
    }
}
