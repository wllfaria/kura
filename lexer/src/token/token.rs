use std::{
    fmt,
    ops::{Range, RangeBounds},
};

use miette::SourceSpan;

use super::kind::Kind;

pub trait IntoToken<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok>;
}

#[derive(Debug, Clone, Copy)]
pub struct Location {
    pub start_byte: usize,
    pub end_byte: usize,
}

impl Location {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start_byte: start,
            end_byte: end,
        }
    }
}

impl From<Location> for SourceSpan {
    fn from(value: Location) -> Self {
        SourceSpan::new(value.start_byte.into(), value.end_byte - value.start_byte)
    }
}

impl From<(usize, usize)> for Location {
    fn from((start_byte, end_byte): (usize, usize)) -> Self {
        Self {
            start_byte,
            end_byte,
        }
    }
}

impl From<Range<usize>> for Location {
    fn from(range: Range<usize>) -> Self {
        let start = match range.start_bound() {
            std::ops::Bound::Included(start) => *start,
            std::ops::Bound::Excluded(start) => *start,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        let end = match range.end_bound() {
            std::ops::Bound::Included(end) => *end,
            std::ops::Bound::Excluded(end) => *end,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        Location {
            start_byte: start,
            end_byte: end,
        }
    }
}

#[derive(Debug)]
pub struct Token<'tok> {
    pub kind: Kind<'tok>,
    pub location: Location,
}

impl<'tok> Token<'tok> {
    pub fn new(kind: Kind<'tok>, location: Location) -> Self {
        Self { kind, location }
    }
}

impl fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.kind)
    }
}
