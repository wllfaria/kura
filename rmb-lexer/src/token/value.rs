use std::fmt;

use super::primitive::Primitive;

#[derive(Debug, PartialEq)]
pub enum Value<'tok> {
    Primitive(Primitive),
    Ident(&'tok str),
    String(&'tok str),
}

impl fmt::Display for Value<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Value::Primitive(p) => write!(f, "{}", p),
            Value::Ident(i) => write!(f, "{i}"),
            Value::String(s) => write!(f, "{s}"),
        }
    }
}
