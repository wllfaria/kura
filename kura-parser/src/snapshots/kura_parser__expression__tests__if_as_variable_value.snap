---
source: kura-parser/src/expression.rs
expression: let_if_ast
---
Var {
    mutable: false,
    name: "my_var",
    typ: None,
    value: If {
        condition: BinaryOp {
            operator: EqualEqual,
            lhs: Ident {
                name: "something",
                location: Location {
                    start_byte: 31,
                    end_byte: 40,
                },
            },
            rhs: Ident {
                name: "another_thing",
                location: Location {
                    start_byte: 44,
                    end_byte: 57,
                },
            },
            location: Location {
                start_byte: 31,
                end_byte: 57,
            },
        },
        location: Location {
            start_byte: 28,
            end_byte: 142,
        },
        truthy: Block {
            expressions: [
                BinaryOp {
                    operator: Plus,
                    lhs: UintLiteral {
                        value: 10,
                        size: None,
                        location: Location {
                            start_byte: 76,
                            end_byte: 78,
                        },
                    },
                    rhs: UintLiteral {
                        value: 10,
                        size: None,
                        location: Location {
                            start_byte: 81,
                            end_byte: 83,
                        },
                    },
                    location: Location {
                        start_byte: 76,
                        end_byte: 83,
                    },
                },
            ],
            location: Location {
                start_byte: 58,
                end_byte: 97,
            },
        },
        falsy: [
            Block {
                expressions: [
                    BinaryOp {
                        operator: Plus,
                        lhs: UintLiteral {
                            value: 20,
                            size: None,
                            location: Location {
                                start_byte: 121,
                                end_byte: 123,
                            },
                        },
                        rhs: UintLiteral {
                            value: 20,
                            size: None,
                            location: Location {
                                start_byte: 126,
                                end_byte: 128,
                            },
                        },
                        location: Location {
                            start_byte: 121,
                            end_byte: 128,
                        },
                    },
                ],
                location: Location {
                    start_byte: 103,
                    end_byte: 142,
                },
            },
        ],
    },
    location: Location {
        start_byte: 13,
        end_byte: 142,
    },
}
